Skip processing plain YAML files marked as template

pivotaljohn opened the pull request:
Addresses issue vmware-tanzu/carvel-ytt#232.

cppforlife commented on pkg/files/file.go:
```diff
@@ -225,14 +228,32 @@ func (r *File) IsTemplate() bool {
+	switch r.Type() {
+	case TypeYAML:
+		for _, ann := range literal.AllAnnotations {
+			if bytes.Contains(data, ann) {
```
i think we are at the wrong abstraction level here. if somebody has a `foo: "#@overlay/match"` content this would pick this up as being a template, even though it is not.

pivotaljohn replied:
I blame the beer. 🍺 

Yeah, this approach was way too literal (pun inadvertent) and subpar in pretty much every measure.

I see how I got fixated on the notion that `File.IsTemplate()` is _the_ source of truth about whether a file contains templating or not. Quite a bit of complexity flowed from that myoptia.

I see now that `files` is about user input: through a combination of filenames and file marks, indicating how they _intend_ a file to be processed.

Whether a given Document Set actually contains templating is squarely in the domain of `yamltemplate`. Leveraging that fact to adjust processing (i.e. skipping processing of files that contain nothing to process) is a decision well-made at the `workspace` level.

cppforlife commented on pkg/files/file.go:
```diff
@@ -225,14 +228,32 @@ func (r *File) IsTemplate() bool {
+			}
+		}
+	case TypeText:
+		return bytes.Contains(data, literal.TextTemplateStartToken)
```
hmm... i am not sure we can do this for text (== always template by default) since there is no such thing as a comment in text.

in the line above, what if user only placed `@)` and screwed up opening tag?

pivotaljohn replied:
... and there's no need to change how text files are processed. This wasn't asked for.

cppforlife commented on pkg/files/file.go:
```diff
@@ -4,11 +4,14 @@
 	"sort"
 	"strings"
+
+	"github.com/k14s/ytt/pkg/literal"
```
feels a bit weird for the lowest layer to pick up annotations from the highest layers.

cppforlife commented on pkg/literal/annotations.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+var AllAnnotations = [...][]byte{
+	[]byte("#@" + CommentAnnotationShorthand),
```
#@ and (@ seems like details that are "owned" by other packages.

cppforlife commented on pkg/cmd/template/cmd.go:
```diff
@@ -60,7 +60,7 @@ func NewCmd(o *Options) *cobra.Command {
 	}
 	cmd.Flags().BoolVar(&o.IgnoreUnknownComments, "ignore-unknown-comments", false,
-		"Configure whether unknown comments are considered as errors (comments that do not start with '#@' or '#!')")
+		"(deprecated: to explicitly disable processing, use `--file-mark <filename>:type=yaml-plain`) Configure whether unknown comments are considered as errors (comments that do not start with '#@' or '#!')")
```
this description is a bit confusing since it almost implies that this flag is deprecated and i should use file marks, even though the main change here is that folks dont need to do anything anymore. this also applies to the --implicit-map-key-overrides flag.

i wouldnt modify this description because it's still true as is for yaml templates.

pivotaljohn replied:
With this detection in place, what's the need for the `--ignore-unknown-comments` flag?

cppforlife replied:
this flag would still be used to ignore unknown comments within legit ytt templates if user wants that.

pivotaljohn replied:
Why would they want this... like what's the use-case that says, "I authored this file, but I still want YAML comments"?

It invites failure states. What convenience does it afford?

cppforlife replied:
i can think of one case: i have a third party yaml file with bunch of comments. i want to start using ytt templating in it so i add one ytt directive. without having to fix all the other comments _right this moment_ i want to start seeing ytt do its evaluation.

jtigger committed:
Skip processing plain YAML files marked as template

cppforlife commented on pkg/workspace/library_loader.go:
```diff
@@ -116,7 +116,8 @@ func (ll *LibraryLoader) filesByAnnotation(annName structmeta.AnnotationName, lo
 
-			tplOpts := yamltemplate.MetasOpts{IgnoreUnknown: ll.templateLoaderOpts.IgnoreUnknownComments}
+			// we're only peeking at the document to determine its "type", no need to detect (yet) if contains unknowns.
+			tplOpts := yamltemplate.MetasOpts{IgnoreUnknown: true}
```
should we move this into DocExtractor?

pivotaljohn replied:
Well played.

Yeah, DocExtractor being insensitive to "unknown comments" in exchange for a simpler interface there feels like a win.

cppforlife commented on pkg/workspace/template_loader.go:
```diff
@@ -160,7 +160,7 @@ func (l *TemplateLoader) EvalYAML(libraryCtx LibraryExecutionContext, file *file
 	docSet.Print(l.ui.DebugWriter())
 
-	if !file.IsTemplate() && !file.IsLibrary() {
+	if !(file.IsTemplate() && yamltemplate.IsTemplate(docSet)) && !file.IsLibrary() {
```
this is getting hard to understand. should we just add a separate if cond with return instead?

pivotaljohn replied:
No so fast, mister... I think we can do it with a little boolean transformation and a more precise name for this new function...

```go
if !file.IsTemplate() && !file.IsLibrary() || !yamltemplate.HasTemplating(docSet) {
```

huh? huh?

cppforlife commented on pkg/yamltemplate/template.go:
```diff
@@ -34,6 +34,33 @@ type TemplateOpts struct {
+
+	metaOpts := MetasOpts{IgnoreUnknown: true}
+	for _, meta := range node.GetMetas() {
+		structMeta, err := NewStructMetaFromMeta(meta, metaOpts)
```
it seems that contract right now for NewStructMetaFromMeta is to return errors for non-valid annotations. if we load it up with `MetasOpts{}` (== `IgnoreUnknown: false`) then we can just check on the error, right?

cppforlife replied:
actually nm, this wouldnt work if there would be no comments, right...

cppforlife replied:
i guess this loop looks a little "annoying" but i dont think we can do much about it.

pivotaljohn replied:
> actually nm, this wouldnt work if there would be no comments, right...

I like your moxy. But, exactly.

> i guess this loop looks a little "annoying" but i dont think we can do much about it.

Yeah, it's also irritatingly identical to some linting logic we do in `DocExtractor`, but the duplication isn't tripping that threshold where the cost of the abstraction would be worth it, yet.

cppforlife commented on pkg/yamltemplate/template.go:
```diff
@@ -34,6 +34,33 @@ type TemplateOpts struct {
 	ImplicitMapKeyOverrides bool
 }
 
+func IsTemplate(val interface{}) bool {
```
it's a little weird to have `val interface{}`. may be have input as DocumentSet and have private isTemplate() function with interface? wdyt?

pivotaljohn replied:
Yeah... I started with that exact structure... but it seemed a bit silly to have an exported function that's a one-liner that calls the unexported function that had all the logic.

Then on top, the first thing we do in the function is downcast that quietly terminates.

From a "when in Rome" perspective, in every other case I saw something like that, here, there was some kind of pre-condition or post-condition work that felt like it justified the separate functions of the same name. 

Perhaps given that it's an exported function is sufficient justification for having as typed of a signature as possible. I'll switch it back. Thanks for asking.

jtigger committed:
Refinements from March 16 review

- make DocExtractor insensitive to "unknown comments", simplifying its
  usage
- rename yamltemplate.IsTemplate() to yamltemplate.HasTemplating(); the
  increase in precision is not only more accurate, but adds a touch of
  clarity to its usage.

cppforlife commented on pkg/workspace/doc_extractor.go:
```diff
@@ -37,6 +36,8 @@ func (v DocExtractor) Extract(annName structmeta.AnnotationName) ([]*yamlmeta.Do
 	return matchedDocs, nonMatchedDocs, nil
 }
 
+var ignoreUnknown = yamltemplate.MetasOpts{IgnoreUnknown: true}
```
should we just inline this? seems a bit weird to have workspace package own this little tidbit as top level var.

pivotaljohn replied:
I mean, we can... I hear you about a package-level dangling var... I guess the fact that we're building the same value object again and again isn't too big a deal.

cppforlife replied:
your call. im ok either way with this.

> the fact that we're building the same value object again and again isn't too big a deal

yup, if we ever start optimizing ytt for perf i dont think this will show up on your radar.

cppforlife commented on pkg/workspace/template_loader.go:
```diff
@@ -160,7 +160,7 @@ func (l *TemplateLoader) EvalYAML(libraryCtx LibraryExecutionContext, file *file
 	docSet.Print(l.ui.DebugWriter())
 
-	if !file.IsTemplate() && !file.IsLibrary() {
+	if !file.IsTemplate() && !file.IsLibrary() || !yamltemplate.HasTemplating(docSet) {
```
i was thinking this reads better:

```
if !file.IsTemplate() && !file.IsLibrary() {
    return nil, docSet, nil
}
if !yamltemplate.HasTemplating(docSet) {
    return nil, docSet, nil
}
```

pivotaljohn replied:
I respectfully disagree.

The way it's written in the PR, it's the same early exit for the same reason and it's all in one sentence:

"If the file is not a template... and it's not a library... Ooooor, it just doesn't have any templating in the first place... well, no need to 'evaluate' it."

cppforlife replied:
ok lets keep it as is.

jtigger committed:
Inline options struct in document extractor

- when we get serious calls for performance needs, we can optimize for
  reducing the number of mallocs. Until then, having a package-level
  variable for a constant like this isn't worth it.
